Add a course setting to not allow self-enrollment

renatho opened the pull request:
Resolves #7182

## Proposed Changes

* Add a course setting to not allow self enrollment.
* Also add notices in the frontend when the setting is enabled.

## Testing Instructions
<!--
Add as many details as possible to help others reproduce the issue and test the changes.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Create a course with blocks and Learning Mode enabled.
2. Enable the setting _"Don't allow self enrollment"_ in the course editor sidebar.
3. Add at least one lesson to the course.
4. As a student, visit the course and the lesson pages.
5. Make sure you see notices and you can't enroll.
6. Enroll the student through the Student Management (WP Admin), and make sure it works properly.
7. Check that you don't see the notices anymore.
8. Repeat the previous steps, but this time with a course without blocks (legacy course), and without Learning Mode.
9. Play around, and make sure there's not take course button for the user when the new setting is enabled.
10. Turn the setting _"Don't allow self enrollment"_ off, and make sure the course works normally (no notices, and it's possible to enroll).

### Known issues

https://github.com/Automattic/sensei/issues/7232

### Screenshots

The setting:
<img width="315" alt="Screenshot 2023-10-18 at 11 40 34" src="https://github.com/Automattic/sensei/assets/876340/cbaed326-1158-4da3-84f4-83655cb01b9a">

Notice on course with blocks:
<img width="1112" alt="Screenshot 2023-10-18 at 14 58 44" src="https://github.com/Automattic/sensei/assets/876340/21c1cfa0-c5f1-44d2-9bc5-a2292d15a3d2">

Notice on lesson with Learning Mode:
<img width="1203" alt="Screenshot 2023-10-18 at 11 41 02" src="https://github.com/Automattic/sensei/assets/876340/acf68e71-aeea-4d84-9998-d03746ca66f9">

Notice on course without blocks (impacted by https://github.com/Automattic/sensei/issues/6173 in some themes):
<img width="1460" alt="Screenshot 2023-10-18 at 12 16 14" src="https://github.com/Automattic/sensei/assets/876340/3fc8fe1e-8a7b-4b9d-83a1-cfecfee1d181">

Notice on lesson without Learning Mode:
<img width="1119" alt="Screenshot 2023-10-17 at 16 42 50" src="https://github.com/Automattic/sensei/assets/876340/19d2e115-afe3-4c74-8499-85012de0c6c0">

Courses list (impacted by https://github.com/Automattic/sensei/issues/7232):
<img width="1192" alt="Screenshot 2023-10-18 at 16 28 17" src="https://github.com/Automattic/sensei/assets/876340/79304ace-f90d-4597-9516-f058f51c00c2">

## New/Updated Hooks
<!-- Add the following only if there are new/updated actions or filters. Please provide a brief description of what they do and any arguments they may take. Be sure to also add the "Hooks" label to this PR. -->

* `sensei_self_enrollment_not_allowed` - Filter whether self enrollment is not allowed for a given course.

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [x] Acceptance criteria is met
- [x] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [x] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [x] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] New UIs match the designs
- [x] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [x] "Needs Documentation" label is added if this change requires updates to documentation
- [x] Known issues are created as new GitHub issues

renatho committed:
Add setting to not allow self enrollment

renatho committed:
Add extra check for self enrollment not allowed on user capability to manually enrol

renatho committed:
Add notice about not allowed self enrollment on lessons not using learning mode

renatho committed:
Add notice about not allowed self enrollment on lessons using learning mode

renatho committed:
Add not allowed self enrollment notice on courses without blocks

renatho committed:
Add since to new method

renatho committed:
Improve test since regex is not needed here

renatho committed:
Add notice to take course button

renatho commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -171,6 +172,24 @@ const CourseGeneralSidebar = () => {
 
+			<HorizontalRule />
+
+			<h3>{ __( 'Enrollment', 'sensei-lms' ) }</h3>
```
It is the setting in the course sidebar.

renatho commented on includes/blocks/class-sensei-block-take-course.php:
```diff
@@ -71,6 +71,12 @@ public function render_take_course_block( $attributes, $content ): string {
 				}
 				$html = $this->render_with_start_course_form( $course_id, $content );
 			}
+		} elseif ( Sensei_Course::is_self_enrollment_not_allowed( $course_id ) && ! Sensei_Course::is_user_enrolled( $course_id, get_current_user_id() ) ) {
```
It is the notice for courses using blocks.

renatho commented on includes/class-sensei-course.php:
```diff
@@ -3728,7 +3742,10 @@ public static function can_current_user_manually_enrol( $course_id ) {
 
-		$default_can_user_manually_enrol = is_user_logged_in() && ! $is_user_enrolled;
+		// Check if self enrollment is not allowed for this course.
+		$is_self_enrollment_not_allowed = self::is_self_enrollment_not_allowed( $course_id );
```
Extra check for this "capability" method, which already solves many cases, like hiding buttons and preventing enrollment.

renatho committed:
Document method as internal since it's only used by a hook

renatho commented on includes/class-sensei-course.php:
```diff
@@ -4098,6 +4128,25 @@ public static function prerequisite_complete_message() {
+	 *
+	 * @internal
+	 */
+	public static function self_enrollment_not_allowed_message() {
```
It adds the notice for courses that don't use blocks.

renatho commented on includes/class-sensei-course.php:
```diff
@@ -4336,6 +4387,9 @@ public function remove_legacy_course_actions() {
 		remove_action( 'sensei_single_course_content_inside_before', [ 'Sensei_Course', 'prerequisite_complete_message' ], 20 );
 
+		// Self enrollment not allowed message.
+		remove_action( 'sensei_single_course_content_inside_before', [ 'Sensei_Course', 'self_enrollment_not_allowed_message' ], 20 );
```
Similar to the other "legacy" hooks (for courses not using blocks), it removes the message, since it's added through the Take Course block.

renatho commented on includes/class-sensei-lesson.php:
```diff
@@ -4782,8 +4782,16 @@ public static function course_signup_link() {
+			$format = '';
+			if ( Sensei_Course::is_self_enrollment_not_allowed( $course_id ) ) {
+				// translators: The placeholder %1$s is a link to the Course.
+				$format = esc_html__( 'Please contact the course administrator to sign up for the %1$s before starting the lesson.', 'sensei-lms' );
```
It adds the notice for lessons where the learning mode is not being used.

donnapep replied:
Could we use the same notice that we use everywhere else (i.e. "Please contact the course administrator to access the course content.") for consistency and to reduce the number of new string that need to be translated?

renatho replied:
Sure thing! Message updated here for consistency: https://github.com/Automattic/sensei/pull/7231/commits/c6ab19a3673a5d65597effb79cbe15a9c4edbad4

renatho commented on includes/course-theme/class-sensei-course-theme-lesson.php:
```diff
@@ -262,6 +262,19 @@ private function maybe_add_not_enrolled_notice() {
 
+		// Check if self enrollment is allowed.
+		if ( Sensei_Course::is_self_enrollment_not_allowed( $course_id ) ) {
+			$notices->add_notice(
```
Lesson notice with Learning Mode.

renatho committed:
Add changelog

renatho committed:
Skip if course ID is false

renatho committed:
Restore initial notices object after tests

renatho committed:
Add filter for self enrollment not allowed

merkushin commented:
Thanks @renatho! Works awesome!

However, it was impossible to enable the option in a non-block course (using Classic Editor). There is no meta-panel for changing it.
Or I got you wrong? How I was supposed to test a legacy course?
At the same time, I think we claim we don't support non-block version anymore?

P.S. Please, take a look at the Psalm 8.2 check ;)

merkushin commented on includes/class-sensei-course.php:
```diff
@@ -3751,6 +3768,32 @@ public static function can_current_user_manually_enrol( $course_id ) {
+		 * @param {int}  $course_id                   Course post ID.
+		 * @return {bool} Filtered value.
+		 */
+		return (bool) apply_filters( 'sensei_self_enrollment_not_allowed', $self_enrollment_not_allowed, $course_id );
```
Just a thought as a potential developer, maybe we can pass the user id here?
In this case, the filter can behave differently depending on the user.
I can imagine scenarios when we might want to allow enrolling for some roles.
Not necessary though. Especially, if we consider it only as a course setting.

renatho replied:
Hm! Thinking a little about this, I'd be a little concerned about adding the user here because the method and hook are not directly tied to the user, but just a course setting. Currently, I think all the things we're doing are tied to users under the hood, but this could not be true in the future. For example, if we have a cron looping through courses and getting this setting to do something.

And if the users want to make a specific check by the current user in some situation anyway, they could just use the `get_current_user_id()`.

WDYT?

merkushin replied:
Sounds reasonable. Thanks for considering.

donnapep commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -51,6 +51,7 @@ const CourseGeneralSidebar = () => {
 	const featured = meta._course_featured;
 	const prerequisite = meta._course_prerequisite;
 	const notification = meta.disable_notification;
+	const senfEnrollmentNotAllowed = meta._self_enrollment_not_allowed;
```
Typo here in the variable name.

renatho replied:
Fixed in https://github.com/Automattic/sensei/pull/7231/commits/5d4190999b77645e56fbeb185631b9da7cfa6cbc

donnapep commented on includes/class-sensei-course.php:
```diff
@@ -668,6 +668,18 @@ public function set_up_meta_fields() {
 		);
+		register_post_meta(
+			'course',
+			'_self_enrollment_not_allowed',
```
We should be prefixing our meta keys as per https://senseip2.wordpress.com/programming-standards-best-practices/naming-conventions/#meta-keys.

Also, we need to delete this meta when we uninstall the plugin.

renatho replied:
@donnapep, do you think it makes sense to also apply this standard to sensei post types? I mean, it's already "encapsulated" under course post type. If yes, I can change this one quickly.
The post type that in an ideal scenario would probably be "sensei_course".

> Also, we need to delete this meta when we uninstall the plugin.

About this, it will automatically get deleted along with the post since we're deleting the posts with the Sensei post types.

donnapep replied:
> do you think it makes sense to also apply this standard to sensei post types?

I feel like we discussed this before but decided against it. Unfortunately I can't recall why. 😐

renatho replied:
No problem! Done in https://github.com/Automattic/sensei/pull/7231/commits/92d2086f91c92c65fb1db20eecfd2a47314f0508

donnapep commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -171,6 +172,24 @@ const CourseGeneralSidebar = () => {
+					} )
+				}
+				help={ __(
+					"Students can't take this course by themselves. They can be enrolled only by teachers, administrators, or invitation.",
```
What about "Students need to be manually enrolled."? I don't think we should mention invitation since that's a Pro feature.

renatho replied:
I used "Students need to be manually enrolled by teachers or administrators.". WDYT?
I thought to use "through WP Admin" instead of "by teachers or administrators", but I thought it could be confusing for invites. https://github.com/Automattic/sensei/pull/7231/commits/32f170d6b2aae214605b7f728489cd87bee8dbe2

donnapep commented on assets/js/admin/course-general-sidebar.js:
```diff
@@ -171,6 +172,24 @@ const CourseGeneralSidebar = () => {
+
+			<h3>{ __( 'Enrollment', 'sensei-lms' ) }</h3>
+			<CheckboxControl
+				label={ __( "Don't allow self enrollment", 'sensei-lms' ) }
```
"self enrollment" has a hyphen "self-enrollment".

renatho replied:
Fixed in https://github.com/Automattic/sensei/pull/7231/commits/dc9500c37976f60c7cc719d99ff2b746a6577878

donnapep reviewed (commented):
Added a couple of things I noticed.

renatho commented:
Hi @merkushin!

> However, it was impossible to enable the option in a non-block course (using Classic Editor). There is no meta-panel for changing it.
Or I got you wrong? How I was supposed to test a legacy course?

You can just remove all the blocks from the course to see the legacy course (keeping it empty on the editor). 😉

> P.S. Please, take a look at the Psalm 8.2 check ;)

Fixed with https://github.com/Automattic/sensei/pull/7231/commits/09b934a8fd1491410656b4c86779217fdd4cfabe

renatho committed:
Fix variable typo

renatho committed:
Update setting help text

renatho committed:
Add hyphen to self-enrollment

renatho committed:
Extract callback to avoid duplicate code

renatho committed:
Merge branch 'trunk' into add/setting-not-allowing-self-enrollment

renatho committed:
Add sensei prefix to meta key

renatho committed:
Add PHPDoc comments

renatho committed:
Fix take course notices being displayed on courses archive page

renatho committed:
Add changelog

renatho committed:
Merge pull request #7236 from Automattic/fix/take-course-notices

Fix course notices that are intended to be displayed only on the course page but were currently appearing on the courses archive page

renatho commented:
Just a note that I accidentally merged https://github.com/Automattic/sensei/pull/7236 which was based on this branch before this is on `trunk`, so feel free to ignore the new changes introduced by https://github.com/Automattic/sensei/pull/7231/commits/e1e02b55fefeecc2d260eb317de50f253ef2d7ca.

donnapep committed:
Merge branch 'trunk' into add/setting-not-allowing-self-enrollment

donnapep reviewed (commented):
Looking good! A wee bit of testing feedback...

For free preview lessons, the content displays as expected. WDYT about removing the notice for preview lessons, since it's a bit misleading, but leaving it on the attached quiz page (if applicable)?

![Screenshot 2023-10-30 at 7 59 46 AM](https://github.com/Automattic/sensei/assets/1190420/1f4a95a8-6e95-407f-8fab-0eed8df522a9)

For legacy courses:
- I see the notice twice for the Course theme. I guess this is a separate issue though. 
- I also see a "or log in to view this course." notice. We should get rid of that notice when self-enrolment isn't allowed, since you won't be able to view it even after logging in.
- There's a "Register" button that doesn't need to be there. I wonder if it would make sense to remove it entirely since they can already click on the (very bad) notice to register?

![Screenshot 2023-10-30 at 8 07 29 AM](https://github.com/Automattic/sensei/assets/1190420/d3ca9884-1649-4885-9479-5b64edd1a05f)

For timed quizzes, the "Start Quiz" button is still available:

![Screenshot 2023-10-30 at 8 41 46 AM](https://github.com/Automattic/sensei/assets/1190420/baaee9d7-39ea-4289-a431-d8a690d6dee2)

donnapep commented:
One more thing I just thought to test. I disabled _Access Permissions_ in Sensei's settings. When I did that, I'm able to see the content (though the notice is still there), but it's probably best that I'm not able to. I think that the _Don't allow self-enrollment_ setting should take precedence, given that it's a more granular check. What are your thoughts?

![Screenshot 2023-10-30 at 8 54 29 AM](https://github.com/Automattic/sensei/assets/1190420/6d65548e-f74f-4793-af97-7cff40a70621)

renatho committed:
Merge branch 'trunk' of github.com:Automattic/sensei into add/setting-not-allowing-self-enrollment

renatho committed:
Unify messages

renatho committed:
Refactor and change preview lesson notice with self-enrollment not allowed

renatho committed:
Don't display register button or log in notice when self enrollment is not allowed

renatho commented:
> For free preview lessons, the content displays as expected. WDYT about removing the notice for preview lessons, since it's a bit misleading, but leaving it on the attached quiz page (if applicable)?

Instead of removing the message completely, I tweaked the text to "take this lesson", similar to the message displayed when it allows self-enrollment and the user is not enrolled yet. WDYT? https://github.com/Automattic/sensei/pull/7231/commits/7cdfb12e98a7d541939fc4f13fcad848c21ce32a

When the lesson preview is enabled:
<img width="1268" alt="Screenshot 2023-10-30 at 14 55 09" src="https://github.com/Automattic/sensei/assets/876340/9dcf2b7c-ba56-4c05-ae86-d0b313a01fcb">

When the lesson preview is not enabled:
<img width="1256" alt="Screenshot 2023-10-30 at 14 55 18" src="https://github.com/Automattic/sensei/assets/876340/63c4630c-e03d-4e5a-924f-3348389f20b8">

> I see the notice twice for the Course theme. I guess this is a separate issue though.

It is! The issue: https://github.com/Automattic/sensei/issues/7237. I tried to fix it as part of https://github.com/Automattic/sensei/pull/7236, but since it wasn't very simple and it's only for legacy courses, I just created the issue for now.

> I also see a "or log in to view this course." notice. We should get rid of that notice when self-enrolment isn't allowed, since you won't be able to view it even after logging in.

> There's a "Register" button that doesn't need to be there. I wonder if it would make sense to remove it entirely since they can already click on the (very bad) notice to register?

Good catch! Fixed in https://github.com/Automattic/sensei/pull/7231/commits/425d44dc5a768bef7d6957676c7ab4b49a2e5b8d
Notice that if the user is logged out, we're not displaying any link to login in any of the messages. I also removed the "or log in to view this course.".

> For timed quizzes, the "Start Quiz" button is still available:

Interesting! I could see it while navigating to the quiz URL directly through the address bar, but I can't see any link to navigate to the quiz. Do we display a link somewhere? Also, I see the same behavior in a normal course when the user is not enrolled.

That said, I'm thinking that we just shouldn't display the quiz at all if the user is not enrolled. 🤔 WDYT? If you agree I'll just open a new issue for that, since it's not directly related to this PR. Also, if we don't have a link to the quiz, I imagine we could consider it's not critical, right?

> One more thing I just thought to test. I disabled Access Permissions in Sensei's settings. When I did that, I'm able to see the content (though the notice is still there), but it's probably best that I'm not able to. I think that the Don't allow self-enrollment setting should take precedence, given that it's a more granular check. What are your thoughts?

Hm! That's a good point! But thinking more, I'm not sure if we should mix the things because the new setting is related to enrollment, while the "Users must be logged in to view lesson content" setting is more related to the content visualization.

And about the messaging, I think the change I did in https://github.com/Automattic/sensei/pull/7231/commits/7cdfb12e98a7d541939fc4f13fcad848c21ce32a helps with that, now saying that they need to contact the administrator to **take the lesson** instead of access the content. WDYT?

donnapep committed:
Merge branch 'trunk' into add/setting-not-allowing-self-enrollment

donnapep commented:
> Do we display a link somewhere?

Nope, I navigated directly to the quiz. 🙂 

> Also, I see the same behavior in a normal course when the user is not enrolled.

Yup, I think we never caught this until now.

> That said, I'm thinking that we just shouldn't display the quiz at all if the user is not enrolled. 🤔 WDYT? If you agree I'll just open a new issue for that, since it's not directly related to this PR.

Not sure about not display the quiz. I guess if they can't see the lesson content then they shouldn't be able to see the quiz content either. TODO: Test this.

donnapep commented:
> Instead of removing the message completely, I tweaked the text to "take this lesson", similar to the message displayed when it allows self-enrollment and the user is not enrolled yet. WDYT?

Sounds good.

> Do we display a link somewhere?

Nope, I navigated directly to the quiz. 🙂 

> Also, I see the same behavior in a normal course when the user is not enrolled.

Yup, I think we never caught this until now.

> That said, I'm thinking that we just shouldn't display the quiz at all if the user is not enrolled. 🤔 WDYT? If you agree I'll just open a new issue for that, since it's not directly related to this PR.

I need to think about this one a bit more and see what we do for legacy courses.

> Also, if we don't have a link to the quiz, I imagine we could consider it's not critical, right?

Yup, agreed! 👍🏻 

> Hm! That's a good point! But thinking more, I'm not sure if we should mix the things because the new setting is related to enrollment, while the "Users must be logged in to view lesson content" setting is more related to the content visualization.
> And about the messaging, I think the change I did in https://github.com/Automattic/sensei/commit/7cdfb12e98a7d541939fc4f13fcad848c21ce32a helps with that, now saying that they need to contact the administrator to take the lesson instead of access the content. WDYT?

Good thought! I suppose if this becomes an issue then we may need to think more about it.
